protocol/x/vault/keeper/withdraw.go

```diff
@@ -332,3 +221,161 @@ func (k Keeper) WithdrawFromMegavault(
 
 	return redeemedQuoteQuantums, nil
 }
+
+// RedeemFromMainAndSubVaults redeems `shares` number of shares from main and sub vaults.
+// If and only if `txMode` is true, logs are enabled and quote quantums redeemed from each
+// sub vault are transferred to the main vault.
+func (k Keeper) RedeemFromMainAndSubVaults(
+	ctx sdk.Context,
+	shares *big.Int,
+	txMode bool,
+) (
+	redeemedQuoteQuantums *big.Int,
+	megavaultEquity *big.Int,
+	totalShares *big.Int,
+	err error,
+) {
+	// Redeem from main vault.
+	totalShares = k.GetTotalShares(ctx).NumShares.BigInt()
+	if shares.Cmp(totalShares) > 0 {
+		return nil, nil, totalShares, errorsmod.Wrapf(
+			types.ErrInvalidSharesToWithdraw,
+			"shares to withdraw %s exceeds total shares %s",
+			shares,
+			totalShares,
+		)
+	}
+	megavaultEquity, err = k.GetSubaccountEquity(ctx, types.MegavaultMainSubaccount)
+	if err != nil {
+		if txMode {
+			log.ErrorLogWithError(ctx, "Megavault withdrawal: failed to get megavault main vault equity", err)
```

Comment:
nit: Perhaps better to log at level debug if it's a query and log at level error if it's in a transaction?